servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java

```diff
@@ -598,21 +712,67 @@ public Completable onClose() {
         @SuppressWarnings("unchecked")
         private Completable doClose(final Function<? super C, Completable> closeFunction) {
             return Completable.defer(() -> {
-                final Object[] connections = connStateUpdater.getAndSet(this, CLOSED_CONN_STATE).connections;
+                final ConnState oldState = connStateUpdater.getAndSet(this, CLOSED_CONN_STATE);
+                cancelIfHealthCheck(oldState.state);
+                final Object[] connections = oldState.connections;
                 return connections.length == 0 ? completed() :
                         from(connections).flatMapCompletableDelayError(conn -> closeFunction.apply((C) conn));
             });
         }
 
+        void markUnhealthy(ConnectionFactory<Addr, ? extends C> connectionFactory, Predicate<C> selector) {
+            if (healthCheckConfig == null) {
+                return;
+            }
+            connectionFailuresUpdater.getAndUpdate(this, val -> val == Integer.MAX_VALUE ? val : val + 1);
+            for (;;) {
+                ConnState previous = connStateUpdater.get(this);
+                if (previous.state != STATE_ACTIVE
+                        || previous.connections.length > 0
+                        || connectionFailuresUpdater.get(this) < this.healthCheckConfig.failedThreshold) {
```

Comment:
Thank you for the suggestion. I considered the race and decided an optimistic assumption that the race only causes lost increments, but still would trigger a health check. I understand it makes sense to avoid any other races and will iterate over this idea. However, you suggest the increments happen on unhealthy state, which previously meant that the background check is already scheduled. Now, the HealthCheck could mean ACTIVE, but with count. For clarity of state, I'd go for creating objects of ACTIVE type with a counter and include your suggestions. Does that seem fair?

Reply:
> Now, the HealthCheck could mean ACTIVE, but with count.

Yes I didn't take into account `failedThreshold` here so that makes sense. Seems like we actually only need the count on the `ACTIVE` state, as when we are unhealthy we don't need to count any more. We should also use an atomic inc/update/cas to compare against `failedThreshold` (instead of two independent operations).